pkg/catalog/catalog.go

```diff
@@ -10,9 +11,140 @@ import (
+	}
+
+	return devfile, nil
+}
+
+// IsDevfileComponentSupported checks if the devfile is supported
+// The supported devfile should satisfy the following conditions:
+// 1. Devfile has dockerimage as component type
+// 2. Devfile has alias
+// 3. Devfile has run command
+// 4. Devfile has build command
+func IsDevfileComponentSupported(devfile Devfile) bool {
+	hasDockerImage := false
+	hasAlias := false
+	hasRunCommand := false
+	hasBuildCommand := false
+
+	for _, component := range devfile.Components {
+		if hasDockerImage && hasAlias {
+			break
+		}
+
+		if !hasDockerImage {
+			hasDockerImage = strings.Contains(component.Type, "dockerimage")
+		}
+
+		if !hasAlias {
+			hasAlias = len(component.Alias) > 0
+		}
+	}
+
+	for _, command := range devfile.Commands {
+		if hasRunCommand && hasBuildCommand {
+			break
+		}
+
+		if !hasRunCommand {
+			hasRunCommand = strings.Contains(command.Name, "run")
+		}
+
+		if !hasBuildCommand {
+			hasBuildCommand = strings.Contains(command.Name, "build")
+		}
+	}
+
+	if hasDockerImage && hasAlias && hasBuildCommand && hasRunCommand {
+		return true
+	} else {
+		return false
+	}
+}
+
+// ListDevfileComponents lists all the available devfile components
+func ListDevfileComponents() (DevfileComponentTypeList, error) {
+	var catalogDevfileList DevfileComponentTypeList
+	catalogDevfileList.DevfileRegistry = DevfileRegistry
+
+	// Load the devfile registry index.json
+	devfileIndex, err := GetDevfileIndex(DevfileIndexLink)
+	if err != nil {
+		return DevfileComponentTypeList{}, err
+	}
+
+	// 1. Load devfiles that indexed in devfile registry index.json
+	// 2. Populate devfile components with devfile data
+	// 3. Form devfile component list
+	for _, devfileIndexEntry := range devfileIndex {
+		devfileIndexEntryName := devfileIndexEntry.DisplayName
+		devfileIndexEntryDescription := devfileIndexEntry.Description
+		devfileIndexEntryLink := devfileIndexEntry.Links.Link
+
+		// Load the devfile
+		devfileLink := DevfileRegistry + devfileIndexEntryLink
+		// TODO: We send http get resquest in this function mutiple times
+		// since devfile registry uses different links to host different devfiles,
+		// this can reduce the performance especially when we load devfiles from
+		// big registry. We may need to rethink and optimize this in the future
+		devfile, err := GetDevfile(devfileLink)
+		if err != nil {
+			return DevfileComponentTypeList{}, err
+		}
+
+		// Populate devfile component with devfile data and form devfile component list
+		var catalogDevfile DevfileComponentType
+		catalogDevfile.Name = strings.TrimSuffix(devfile.MetaData.GenerateName, "-")
+		catalogDevfile.DisplayName = devfileIndexEntryName
+		catalogDevfile.Description = devfileIndexEntryDescription
+		catalogDevfile.Link = devfileIndexEntryLink
+		catalogDevfile.Support = IsDevfileComponentSupported(devfile)
+		catalogDevfileList.Items = append(catalogDevfileList.Items, catalogDevfile)
```

Comment:
All of this can be refactored like so:

```go
catalogDevFile := DevFileComponentType{
 Name: strings.TrimSuffix(devfile.MetaData.GenerateName, "-"),
 DisplayName: devfileIndexEntryName,
...
}
```